11ty site conversion + redesign

Prospero23 opened the pull request:
Slowly have been trying to learn 11ty better and convert the example site into it. 

Created 2/3 column design with optional table of contents automatically generated by headers on page for longer examples. Design is based off of vexflow-demos and liquid docs as mentioned in #13 

CSS files are slightly messy for now, but have adopted the color scheme from vexflow-docs. 

Starting this PR just to show that progress is being made. My goal after the literal conversion is to slowly create a better structure similar to what @ronyeh  has done to allow toggling examples for different vexflow versions. 

No new examples are being added in this PR. 

A deployed version can be found at : https://prospero23.github.io/vexflow-examples/

Prospero23 committed:
basic 11ty structure created (no github actions to deploy yet)

Prospero23 committed:
basic three col layout scafold

Prospero23 committed:
working page-toc

Prospero23 committed:
basic highlight of toc as page scrolls down

Prospero23 committed:
added smooth scroll and basic ol/active styles

Prospero23 committed:
entry examples to basic template 11ty

Prospero23 committed:
highlight js layout conditional added

Prospero23 committed:
basic font 11ty integration

Prospero23 committed:
modules now have basic 11ty integration

Prospero23 committed:
added simpler 2 col layout option with no toc

Prospero23 committed:
active link for whatever page currently on

Prospero23 committed:
added basic 11ty opentype and worker

Prospero23 committed:
added node examples to 11ty

Prospero23 committed:
added node examples to _includes

Prospero23 committed:
added component html tutorials

Prospero23 committed:
added interaction

Prospero23 committed:
added github build step

Prospero23 committed:
small style fixes and better 1 sidebar layout

Prospero23 committed:
basic highlight js dark theme using media attribute on layout

Prospero23 committed:
added button to js and optional space for extra links/styles in header

Prospero23 committed:
made global layout and renamed layout.html to base.html

mscuthbert commented:
Hi Ben -- because you've replaced one file type with another, it's impossible for the maintainers of the repo to see what you are doing on any individual file -- since this PR hasn't been discussed or approved, I would suggest that you create it again with a minimum number of file changes showing what your tool is changing on each file.  Otherwise we'll never be able to verify that every one of the 100+ files has been preserved or approved.

Or initiate a discussion in the appropriate forum.

I am very big on accessbiility improvements but I am big -1 on merging this because I cannot verify that there isn't anything broken.  Security recommendation is no merge.

mscuthbert commented:
Most of the committed .html files do not validate as HTML.

ronyeh commented:
Hi @mscuthbert !

Sorry we never had a chance to introduce you to Ben / @Prospero23 ...

Ben has joined us to help out on the web side of things. In the vexflow.github.io repo, he has created a basic structure for the future vexflow.com. As of now, it is an outer shell that wraps three separate websites, examples, docs, and sandbox. 

We decided in an earlier discussion (with Rodrigo) that we can build this site with a static website generator called 11ty.

11ty uses templates that are processed by the liquid template engine. HTML files and MD files can use includes and other special tags... so it's possible they won't validate as HTML/MD/etc. 

I would recommend that Ben continue with this PR and get to a point that he's happy with the result. He can also clean up the commit history via squash or rebase. Then we can review and test it out and merge. 

If you have any more concerns, please let us know. It's a big undertaking but I think we will get to a good end result faster if we trust Ben and then evaluate the final product. 

I saw you have a PR about vf vs factory. I'll go ahead and review & merge, and then Ben can include those changes into his PR. 

Right now, I'm assuming that no one else is working in this repo. (I'm in the process of improving the docs repo.) @mscuthbert  let me know if you have any other PRs coming so we don't step on each other's toes. I'm assuming you're mostly working in the main vexflow repo on formatter issues. 

If you want, maybe we can schedule a quick zoom/google meet with Ben, Rodrigo, and I just to say hi?

Prospero23 commented:
Hey @mscuthbert! 

I am happy to do whatever makes everyone the most comfortable. Since it got closed, probably worth explicitly noting here that the discussion that sparked starting this process was #11. 

## File Extension Notes
>Most of the committed .html files do not validate as HTML.

When working on the main site repo, I originally marked files with that non-valid syntax as *.liquid but @ronyeh suggested using *.md and *.html as a way to have better syntax highlighting and pretty print in editors (https://github.com/vexflow/vexflow.github.io/issues/1). 

I have tried to do my best keeping file extensions constant during this obviously large conversion, but here is a quick attempt of summing up why files are as they are. 

### html
At the moment, the *.html files are a mix of the key parts of files I originally moved over from the main vexflow repo in #10 and 11ty layouts (node-example, html-example, and base). These layouts still need to be improved, both in naming and API, but they already do a good job of minimizing the amount of styling/repetition needed for creating new examples. 

### md
The markdown files are those that were originally placed in this repo by @rvilarl and I along with a few files just in the project to allow for auto-generation of the navigation structure. These example files were using markdown-it and good old jquery to convert a example.html and example.md page into one index.html page inside of the example/ directory. 11ty makes it easier for contributors not have to think about the internals of these conversions and just simply create new examples. 
## Meeting 
I am in Boston until Wednesday this week, but I will be home and free whenever after that :)

This PR is definitely not complete yet, and I am planning on writing something more substantial to explain all the design decisions etc when everything is in a more final form. I would be happy to discuss everything in a meeting of some sort as well, if that is what would be better for everyone. 

All in all, just want to help out and not step on any toes!

ronyeh commented:
Here was some more discussion about the general website design:
https://github.com/vexflow/vexflow.github.io/issues/1

I did suggest using `.html` and `.md` extensions for the [liquid templates](https://liquidjs.com/tutorials/intro-to-liquid.html). I think it's okay that the template  files don't validate properly as HTML or MD. It allows maintainers to quickly see which file type it "should" be, and many text editors will try their best to syntax highlight and pretty-print the template file.

Let us know of any additional questions / concerns you might have. I gave Ben the go ahead to explore because we need all the help we can get, and he seemed eager to help. 😊 So as of now, @Prospero23 is our lead web site developer, hehe. Please send any requests his way 😉.

mscuthbert commented:
AH! Apologies @Prospero23 — I didn't have any idea that this was under active discussion -- my bad!  THANK YOU and move ahead.  :-)

Prospero23 committed:
restructured css files to be more modular.

Prospero23 committed:
added output styling and restyled loadhtml. Fixed small style bugs in other files

Prospero23 committed:
improved site-nav and page-nav styles

Prospero23 committed:
modularized eleventy config modules

Prospero23 committed:
added copy button transform. cleaned up styles, removed old copy buttons, proper sizing on pre elements

Prospero23 committed:
improved button look

Prospero23 committed:
fixed typo in canvas-interaction example

Prospero23 committed:
new css for opentype/canvas

Prospero23 committed:
changed bg color in worker demo

Prospero23 committed:
merged html-example and node-example into one layout named simple-example

Prospero23 committed:
deleted original test page and cleaned up home page

Prospero23 committed:
added demo shortcode and highlight on build.

Prospero23 committed:
moved js to main.js and added glyph inspector back

Prospero23 committed:
added unit tests back in

Prospero23 committed:
added basic light/dark for prism highlighting

Prospero23 committed:
fixed mjs highlighting and changed demos to all be html

Prospero23 committed:
fix link addresses in module and script-tag

Prospero23 committed:
simplified css. removed page toc and related plugins. small bug fixes

Prospero23 committed:
added tutorial and getting started to guides. small bug fixes.

Prospero23 committed:
change _includes/code to /examples and fix in files

Prospero23 committed:
added global data file and ver shortcode

Prospero23 committed:
added correct order nums for nav

Prospero23 committed:
added font-load-1 example back

Prospero23 committed:
remove old font-load example

Prospero23 committed:
added documentation

Prospero23 committed:
updated README

Prospero23 committed:
added necessary eleventyComputed and cleaned up css in examples

Prospero23 commented:
I think this is pretty much ready for review.... A couple of light mode colors should be changed probably but I would like to get other's opinions first. 

# PR notes
This PR aims to convert the current vexflow-examples site into using Eleventy.  Below I have listed my rational for decisions, organized by directory. The overall structure of the repo is modeled after the vexflow-sandbox, and deviations are explained as needed.

A working preview of this PR can be found here: https://prospero23.github.io/vexflow-examples/

## Lib
This directory is not present in vexflow-sandbox. It stores all of the plugin/shortcode/transform functions and their setups. This feels more modular and maintainable to me then just storing them in the .eleventy.js file. 

### plugins
The EleventyHtmlBasePlugin is used exactly like how it is in vexflow-sandbox.

I am using the 11ty navigation plugin to create the navigation tree for the site. Documentation: https://www.11ty.dev/docs/plugins/navigation/. To allow for programatic building of the nav, I have placeholder files in each of the demos/ directories to act as the “parent” for whatever other demos are also in that folder. To add an example to the navigation, you simply have to add:  

```
eleventyNavigation:
	key: whatever
	parent: parent name
	title: new example
```

I replaced highlight.js with the official 11ty highlighting plugin. Using client-side highlighting was causing pages to momentarily flash unhighlighted code whenever loading a new page. This plugin highlights on build, so its fixes this problem. The documentation can be found here: https://www.11ty.dev/docs/plugins/syntaxhighlight/. It has a good number of configuration options if we eventually need more customization. 

### Shortcodes
I made a demo shortcode to allow for easily creating new examples without having to think about iframes/highlighting/jsfiddle/button creation. Simply just call {% demo params… %} and all of that will be handled for you. There are currently options for highlighting language, showing “run code” button, showing iframe/jsfiddle link, and height of the iframe rendered.

```
{% demo "js" %}
console.log("wow what a demo")
{% enddemo %}
```

I migrated all the examples that had `<pre><code>` blocks to use this new demo shortcode. 

The only annoying thing for the moment with this is that for now, liquid shortcodes only allow for positional arguments. Details about this are in https://github.com/11ty/eleventy/issues/2679. It is possible though that in the near future, changing the parser to “builtin” for liquid will actually work to allow for named arguments and using the shortcode will be even easier. 

 A small usage snag with the shortcode is that the language must be in quotes to work (”js” not js). No quotes = param gets interpreted as a variable. 

The other version shortcode just makes it easy to get whatever the current version of the VexFlow API is to feed into links without having to manually change them. You can simply just say 
```
{% ver v5 %}
```
and it will give you 5.0.0.

### transforms
The only transform I added creates a copy button for every `<pre><code>` block. Seems nice to these elements automatically this without users having to worry about making their own. 

## src
Demo examples are html . It may make sense to formally switch some files over to liquid (especially files like src/_includes/layouts/simple-example.html )for better syntax highlighting. Maybe it makes sense to have layout files be *.liquid and the examples themselves be all *.html. Using *.html to me instead of *.md emphasizes keeping examples from turning into large doc pages.

### _data
Directory functions the same as vexflow-sandbox. I went with global.js instead of .json to allow for storing the current version of the VexFlow api as a variable for easier updating. layout.js makes all the pages on the site default to using the base layout mentioned in the next section. 

### _includes (examples, partials, layouts)
examples/ functions the same as vexflow-sandbox but I have just split files into html and node (js). 

With layouts, base.html is used as the core for all pages on the site, while simple-example.html allows an easy way to make small examples.

### demos
All demo examples are *.html. They all use {% ver %} shortcode where possible for easier maintenance. 

Note: whenever the {% ver %} shortcode is used in the extraHead front matter key, extraHead must be wrapped in eleventyComputed to work correctly (see examples like entry/vexflow-bravura)

### guides
Holds the two longer examples. Uses .md as the files are more text heavy. The one trade-off with using markdown is that since ‘markdown-it’ is used by eleventy to convert markdown into html, I can’t seem to get my shortcode to work with examples that have spaces in them. Once spaces are added between lines in an example, markdown-it converts the code being spit into the iframe as `<p>` tags and refuses to work. There may be a setting to change this behavior, but I have not found it yet. 

## static
Houses slightly different directory names than sandbox (js vs j and css vs c). I favor these longer names, but I am fine switching to the shorter as well. It makes sense to keep everything consistent, whatever we chose. 

### Theme/CSS
I stole the theme colors from vexflow-docs. They work for me, at least for now. Gives us both light and dark modes. 

I have css split into different files separated by function to make development easier. It would be easy enough to bundle the five if we want to.

Prospero23 commented:
I may have missed something in the notes, but am happy to explain anything. I know this PR is massive, so let me know if there is anything I can do to make it more manageable to look over.

Prospero23 committed:
make navigation bg colors clearer

Prospero23 commented:
last small fix to the css I think makes everything look passable.

rvilarl commented:
@Prospero23 it looks really good, thank you!

ronyeh commented on lib/shortcodes/demo.js:
```diff
@@ -0,0 +1,67 @@
+           `
+        : "";
+         
+    return `
```
This line looked like a bug to me at first.

ronyeh commented on src/_includes/examples/html/canvas-interaction.html:
```diff
@@ -5,10 +5,10 @@
   <body>
     <!-- Canvas where the scores will be output -->
-    <canvas id="output"></div>
+    <canvas id="output"></canvas>
```
Nice catch.

ronyeh commented on src/_includes/layouts/base.html:
```diff
@@ -0,0 +1,50 @@
+<body>
+   <div class="container">
+      {% include "partials/nav.html" %}
+      
```
is this extra space required because partials/nav.html would be included above and break the formatting?

ronyeh commented on src/demos/node/customcontext.html:
```diff
@@ -0,0 +1,10 @@
+filename: node/customcontext.js
+
+eleventyNavigation:
+    key: customContext
```
This comment refers to my comment below about keys: Import1, Import2, etc...

ronyeh commented on src/demos/modules/script-tag.html:
```diff
@@ -0,0 +1,44 @@
+title: Module and Script
+
+eleventyNavigation:
+    key: Module and Script
```
Here the key is "Module and Script". So I guess it doesn't really matter as long as Key is a unique string. But maybe we should standardize to a camelCaseKey.

ronyeh reviewed (approved):
Thank you so much! This was a huge undertaking. I have some comments that I'm writing up that I'll append to this discussion, but I'm going to go ahead and merge this now. We can go back and revisit the bugs / comments later.

Prospero23 commented:
Whenever you get a chance, we just need to make sure to switch the deployed branch over to the new gh-pages one instead of main :)
